metrics/metrics.go

```diff
@@ -0,0 +1,202 @@
+	case http.StateNew:
+		open = atomic.AddInt64(&h.openConnections, 1)
+	case http.StateHijacked:
+		fallthrough
+	case http.StateClosed:
+		open = atomic.AddInt64(&h.openConnections, -1)
+	default:
+		return
+	}
+	h.stats.Gauge(fmt.Sprintf("%s.HTTP.OpenConnections", h.statsPrefix), open, 1.0)
+}
+
+// Handle wraps handlers and records various metrics about requests to these handlers
+// and sends them to StatsD
+func (h *HTTPMonitor) Handle() http.Handler {
+	return http.HandlerFunc(h.watchAndServe)
+}
+
+func (h *HTTPMonitor) watchAndServe(w http.ResponseWriter, r *http.Request) {
+	h.stats.Inc(fmt.Sprintf("%s.HTTP.Rate", h.statsPrefix), 1, 1.0)
+	inFlight := atomic.AddInt64(&h.connectionsInFlight, 1)
+	h.stats.Gauge(fmt.Sprintf("%s.HTTP.OpenConnections", h.statsPrefix), inFlight, 1.0)
+
+	cOpened := time.Now()
+	h.handler.ServeHTTP(w, r)
+	cClosed := time.Since(cOpened)
+
+	inFlight = atomic.AddInt64(&h.connectionsInFlight, -1)
+	h.stats.Gauge(fmt.Sprintf("%s.HTTP.ConnectionsInFlight", h.statsPrefix), inFlight, 1.0)
+
+	endpoint := ""
+	// If request fails don't record the endpoint as an attacker could use this to
+	// eat up all our memory by just hitting 404s all day
+	if w.Header().Get("Content-Type") == "application/problem+json" {
+		endpoint = "Failed"
+	} else {
+		// If r.URL has more than two segments throw the rest away to simplify metrics
+		segments := strings.Split(r.URL.Path, "/")
+		if len(segments) > 3 {
+			segments = segments[:3]
+		}
+		endpoint = strings.Join(segments, "/")
+	}
+	h.stats.TimingDuration(fmt.Sprintf("%s.HTTP.ResponseTime.%s", h.statsPrefix, endpoint), cClosed, 1.0)
+}
+
+// RPCMonitor stores rpc delivery state
+type RPCMonitor struct {
+	deliveryTimings map[string]time.Time
+	dtMu            *sync.RWMutex
+
+	stats statsd.Statter
+	clock clock.Clock
+}
+
+// NewRPCMonitor returns a new initialized RPCMonitor and starts a goroutine
+// to cleanup timeouts from the delivery map
+func NewRPCMonitor(stats statsd.Statter) RPCMonitor {
+	r := RPCMonitor{
+		clock:           clock.Default(),
+		stats:           stats,
+		deliveryTimings: make(map[string]time.Time),
+		dtMu:            &sync.RWMutex{},
+	}
+	go func() {
+		c := time.Tick(time.Second * 5)
+		for _ = range c {
+			if t := r.cleanup(); t > 0 {
+				stats.Inc("RPC.Timeouts", t, 1.0)
+			}
+		}
+	}()
+	return r
+}
+
+func (r *RPCMonitor) size() int {
+	r.dtMu.RLock()
+	defer r.dtMu.RUnlock()
+	return len(r.deliveryTimings)
+}
+
+func (r *RPCMonitor) get(id string) (time.Time, bool) {
+	r.dtMu.RLock()
+	defer r.dtMu.RUnlock()
+	timing, present := r.deliveryTimings[id]
+	return timing, present
+}
+
+func (r *RPCMonitor) add(id string) {
+	now := r.clock.Now()
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	r.deliveryTimings[id] = now
+}
+
+func (r *RPCMonitor) delete(id string) {
+	r.dtMu.Lock()
+	defer r.dtMu.Unlock()
+	delete(r.deliveryTimings, id)
+}
+
+func (r *RPCMonitor) cleanup() (removed int64) {
+	checkTime := r.clock.Now().Add(-time.Second * 10)
```

Comment:
This is based on the hard timeout in [`amqp-rpc.go`](https://github.com/letsencrypt/boulder/blob/master/rpc/amqp-rpc.go#L581) which is set to 10 seconds, so any calls without responses called more than 10 seconds ago must have timed out.